### ISSUE TITLE
A Simple question about how to save a model and use it

### ISSUE THREAD

**Issue by hanfeng0409** [NervanaSystems/distiller#183]
I use yaml file in the agp and can get a sparsity 50% on my own resnet, i see there are many zero in model weights, but the size of model saved by distiller was same with before, does it need some other method to do about save model?

**Comment by hanfeng0409** [comment]
I use yaml file in the agp and can get a sparsity 50% on my own resnet, i see there are many zero in model weights, but the size of model saved by distiller was same with before, does it need some other method to do about save model?

**Comment by nzmora** [comment]
Hi @hanfeng0409 ,

When you perform fine-grained pruning (as I assume you did; but note that we extended the basic AGP to support also filter/channel pruning), the weight tensors are not reduced in size because the zero-coefficients are still present in the tensor.
See [Exploring the Regularity of Sparse Structure in Convolutional Neural Networks](https://arxiv.org/pdf/1705.08922.pdf) for an exploration of the different types of sparsities.
![image](https://user-images.githubusercontent.com/31280975/53942464-807d3380-40c3-11e9-8593-ff6ad1e6eba9.png)

Cheers,
Neta

**Comment by hanfeng0409** [comment]
Thank you very much @nzmora ,
and  i will read the paper to learn more. For that i'm new about compression and distiller, i don't know if i need to reduce the size of weights by write code myself or use tools from distiller?
:)

**Comment by akurniawan** [comment]
Hi @nzmora, I also have the same question with @hanfeng0409. After retraining my LSTM net with AGP scheduler, 70% of my weights are now zeroes. However, I don't seem to feel the benefit of this sparsity since the inference latency and the model size are still the same. Do we need to do something else after the pruning part?

**Comment by nzmora** [comment]
Hi @akurniawan,

I've added some more information on this topic in a new [FAQ](https://github.com/NervanaSystems/distiller/wiki/Frequently-Asked-Questions-(FAQ)) page in our Wiki.  This question came up several times and deserves a place in the FAQ.  
Your question regarding quantization was also answered in some older issues, and we will add a detailed answer to the FAQ.
Cheers
Neta

**Comment by RizhaoCai** [comment]
> Hi @akurniawan,
> 
> I've added some more information on this topic in a new [FAQ](https://github.com/NervanaSystems/distiller/wiki/Frequently-Asked-Questions-(FAQ)) page in our Wiki. This question came up several times and deserves a place in the FAQ.
> Your question regarding quantization was also answered in some older issues, and we will add a detailed answer to the FAQ.
> Cheers
> Neta

Hello.

For some pruning methods, the weights would be sparse and the zero-value elements are still present in the tensors (the sizes of weights are not changed). 

However, for some channel pruning methods, the size of weights of conv layers would be changed after pruning.  I still don't know how to save a pruned model and load it.

For example, I define a conv2d layer with in_channels=3, out_channels=64 and kernel_size=(3,3). If I conduct a filter pruning. The out_channels would be reduced to be 52 (assumed). Therefore, the new size of the weight of this conv layer is (52, 3, 3, 3).

Therefore, the weight loaded from the saved checkpoint should be of the same shape (52, 3, 3, 3), isn't it? However, how could I define the Model that could load this new state_dict? If I define the old model, where the shape of the weight is (64, 3, 3, 3), the model cannot load this state_dict, right? But I haven't seen any information that could guide me to define a model that could fit the pruned state_dict

**Comment by nzmora** [comment]
Hi @RizhaoCai ,

You do not need to do anything explicitly when loading the model - it's all taken care of in the background.  However, after pruning you need to add a "thinning" step which will actually remove filters/channels from the weights tensors.  I've added some information to the [FAQ](https://github.com/NervanaSystems/distiller/wiki/Frequently-Asked-Questions-(FAQ)), please see Pruning Q4.

Hope this helps,
Neta

**Comment by RizhaoCai** [comment]
> Hi @RizhaoCai ,
> 
> You do not need to do anything explicitly when loading the model - it's all taken care of in the background. However, after pruning you need to add a "thinning" step which will actually remove filters/channels from the weights tensors. I've added some information to the [FAQ](https://github.com/NervanaSystems/distiller/wiki/Frequently-Asked-Questions-(FAQ)), please see Pruning Q4.
> 
> Hope this helps,
> Neta

Thank you very much. Very helpful

**Comment by akurniawan** [comment]
Hi @nzmora , is there any limitation for the network thinner? such as what sort of layers are you currently supporting? only cnn or lstm is supported as well? 

I also notice in the configuration you have `arch: 'resnet20_cifar'`, does this mean currently the thinner network only supports certain architecture?

**Comment by andyqian2015** [comment]
@RizhaoCai has you encounter this situation that the NNZ drop from 223536 to 120000 but the GPU memory used just drop from 1023M to 985M, just drop a little，why

**Comment by nzmora** [comment]
Closing due to inactivity. Please reopen if needed.
